ci_group/revolve2/ci_group/morphological_novelty_metric/_coordinate_operations.py

```diff
@@ -0,0 +1,116 @@
+from dataclasses import dataclass
+from itertools import product
+
+import numpy as np
+from numpy.typing import NDArray
+
+from revolve2.modular_robot.body import Module
+from revolve2.modular_robot.body.base import Body
+
+
+@dataclass(init=False)
+class CoordinateOperations:
+    """Transform points in a distribution."""
+
+    _coords: list[NDArray[np.float128]]
+
+    def coords_from_bodies(
+        self, bodies: list[Body], cob_heuristics: bool
+    ) -> list[NDArray[np.float128]]:
+        """
+        Extract coordinates of modules from a body.
+
+        :param bodies: The bodies.
+        :param cob_heuristics: If change of basis heuristic approximation is used.
+        :return: The array of coordinates.
+        """
+        self._body_to_adjusted_coordinates(bodies)
+        if cob_heuristics:
+            self._coordinates_pca_heuristic()
+        else:
+            self._coordinates_pca_change_basis()
+
+        try:
+            return self._coords
+        finally:
+            """Ensure new operations are independent from old ones."""
+            del self._coords
```

Comment:
I would say doing this is smelly code and takes long for a reader to understand what is going on. Furthermore, your class is listed as having the member _coords. this means that after the __init__ that variable should always be present. here you break that promise. Consider making this class a function and passing _coords as an argument.

Reply:
i restructured the code for this class so it is just a function, without the dataclass around it. Also removed the del, since its not needed anymore.